Add new measurement PodPeriodicCommand

learnitall opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This commit adds a new measurement named PodPeriodicCommand, which allows users to run commands within pods on a periodic basis, saving the output for later use. For instance, this can be used to scrape pprofs from a select set of pods. Here is an example configuration:

```yaml
name: testppc
namespace:
  number: 1
steps:
- name: Start PodPeriodicCommand
  measurements:
  - identifier: PodPeriodicCommand
    method: PodPeriodicCommand
    params:
      action: start
      labelSelector: k8s-app=cilium
      interval: 30s
      container: cilium-agent
      limit: 1
      failOnError: true
      commands:
      - name: AllocsProfile
        command:
        - curl
        - -s
        - -k
        - http://localhost:6060/debug/pprof/allocs?seconds=30
        timeout: 35s
      - name: Example Timeout
        command:
        - sleep
        - 10
        timeout: 1s
- name: Sleep for two minutes
  measurements:
  - Identifier: ExecCommand
    Method: Exec
    Params:
      timeout: 121s
      command:
      - sleep
      - 120
- name: Gather PodPeriodicCommand
  measurements:
  - identifier: PodPeriodicCommand
    method: PodPeriodicCommand
    params:
      action: gather
```

Running `go tool pprof -http :8080 *AllocsProfile*` in the results directory will open up a new browser tab containing a view of the allocs profiles collected during the two minute run.

Example stats summary using the above config:

```json
{
  "execs": 10,
  "nonZeroRCs": 0,
  "errors": 0,
  "timeouts": 5,
  "measurements": 7,
  "ticks": 5,
  "ticksNoPods": 0
}
```

Example report summary using the above config:

```json
[
  {
    "pod": "cilium-hplhd",
    "namespace": "kube-system",
    "container": "cilium-agent",
    "commands": [
      {
        "exitCode": 0,
        "name": "Example Timeout",
        "command": [
          "sleep",
          "10"
        ],
        "timeout": "1s",
        "hitTimeout": true,
        "startTime": "2023-09-01T12:36:48.762289489-06:00",
        "endTime": "2023-09-01T12:36:49.762788275-06:00"
      },
      {
        "exitCode": 0,
        "name": "AllocsProfile",
        "command": [
          "curl",
          "-s",
          "-k",
          "http://localhost:6060/debug/pprof/allocs?seconds=30"
        ],
        "timeout": "35s",
        "hitTimeout": false,
        "startTime": "2023-09-01T12:36:48.762298737-06:00",
        "endTime": "2023-09-01T12:37:19.079579364-06:00"
      }
    ]
  },
  {
    "pod": "cilium-hplhd",
    "namespace": "kube-system",
    "container": "cilium-agent",
    "commands": [
      {
        "exitCode": 0,
        "name": "Example Timeout",
        "command": [
          "sleep",
          "10"
        ],
        "timeout": "1s",
        "hitTimeout": true,
        "startTime": "2023-09-01T12:37:19.080434913-06:00",
        "endTime": "2023-09-01T12:37:20.081270277-06:00"
      },
      {
        "exitCode": 0,
        "name": "AllocsProfile",
        "command": [
          "curl",
          "-s",
          "-k",
          "http://localhost:6060/debug/pprof/allocs?seconds=30"
        ],
        "timeout": "35s",
        "hitTimeout": false,
        "startTime": "2023-09-01T12:37:19.080752313-06:00",
        "endTime": "2023-09-01T12:37:49.391037823-06:00"
      }
    ]
  },
  {
    "pod": "cilium-hplhd",
    "namespace": "kube-system",
    "container": "cilium-agent",
    "commands": [
      {
        "exitCode": 0,
        "name": "Example Timeout",
        "command": [
          "sleep",
          "10"
        ],
        "timeout": "1s",
        "hitTimeout": true,
        "startTime": "2023-09-01T12:37:49.392397474-06:00",
        "endTime": "2023-09-01T12:37:50.392593022-06:00"
      },
      {
        "exitCode": 0,
        "name": "AllocsProfile",
        "command": [
          "curl",
          "-s",
          "-k",
          "http://localhost:6060/debug/pprof/allocs?seconds=30"
        ],
        "timeout": "35s",
        "hitTimeout": false,
        "startTime": "2023-09-01T12:37:49.392824859-06:00",
        "endTime": "2023-09-01T12:38:19.748587964-06:00"
      }
    ]
  },
  {
    "pod": "cilium-hplhd",
    "namespace": "kube-system",
    "container": "cilium-agent",
    "commands": [
      {
        "exitCode": 0,
        "name": "Example Timeout",
        "command": [
          "sleep",
          "10"
        ],
        "timeout": "1s",
        "hitTimeout": true,
        "startTime": "2023-09-01T12:38:19.749415287-06:00",
        "endTime": "2023-09-01T12:38:20.749768968-06:00"
      },
      {
        "exitCode": 0,
        "name": "AllocsProfile",
        "command": [
          "curl",
          "-s",
          "-k",
          "http://localhost:6060/debug/pprof/allocs?seconds=30"
        ],
        "timeout": "35s",
        "hitTimeout": false,
        "startTime": "2023-09-01T12:38:19.749570704-06:00",
        "endTime": "2023-09-01T12:38:50.576227327-06:00"
      }
    ]
  },
  {
    "pod": "cilium-swbdc",
    "namespace": "kube-system",
    "container": "cilium-agent",
    "commands": [
      {
        "exitCode": 0,
        "name": "Example Timeout",
        "command": [
          "sleep",
          "10"
        ],
        "timeout": "1s",
        "hitTimeout": true,
        "startTime": "2023-09-01T12:38:50.577293404-06:00",
        "endTime": "2023-09-01T12:38:51.577569606-06:00"
      },
      {
        "exitCode": 0,
        "name": "AllocsProfile",
        "command": [
          "curl",
          "-s",
          "-k",
          "http://localhost:6060/debug/pprof/allocs?seconds=30"
        ],
        "timeout": "35s",
        "hitTimeout": false,
        "startTime": "2023-09-01T12:38:50.577536494-06:00",
        "endTime": "2023-09-01T12:39:21.794034396-06:00"
      }
    ]
  }
]
``` 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes: #2262

#### Special notes for your reviewer:

Finishing this work went a bit stale due to some changing priorities, but an initial implementation was used to collect data for the following PRs:

* https://github.com/cilium/cilium/pull/26238
* https://github.com/cilium/cilium/pull/26327

learnitall commented:
CC: @marseel

marseel commented:
/ok-to-test

learnitall commented:
Hey @marseel, @wojtek-t and @mborsz, what would be the next steps here?

marseel commented on clusterloader2/pkg/measurement/common/pod_command.go:
```diff
@@ -0,0 +1,697 @@
+		return false
+	}
+
+	for _, c := range pod.Status.ContainerStatuses {
```
We can probably get rid of checking container status as `PodReady` condition checks for all containers to be in a ready state: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-status

learnitall replied:
Oh great! TIL. Removed.

marseel commented on clusterloader2/pkg/measurement/common/pod_command.go:
```diff
@@ -0,0 +1,697 @@
+		default:
+			return nil, err
+		}
+	case <-t.C:
```
We could simplify it a bit with `time.After(timeout)` and get rid of timer.

learnitall replied:
I'm not sure if we could. This code checks if a timeout occurs, but it is also structured so that the function can return early when this happens. If we use `time.After(timeout)` and remove the timer, then I don't think we would be able to return as soon as a timeout occurs. I think we could only check if a command took longer than the timeout after the command has finished on its own.

marseel commented on clusterloader2/pkg/measurement/common/pod_command.go:
```diff
@@ -0,0 +1,697 @@
+		"%s: finished running commands on pod %s/%s", podPeriodicCommandMeasurementName, pod.Namespace, pod.Name,
+	)
+
+	if len(resultCh) == 0 {
```
I am wondering if it would make sense to skip this check and also plumb `err` from `runCommandInPod` to `runCommandResult`. That would give us a few improvements:
- we would see that command were executed in results
- we would get all errors + we wouldn't need to use `errOnce`
- we could return `runAllCommandsResult` in this method and append it in `commandWorker`.
What do you think?

learnitall replied:
This makes sense! I finished making these changes and will test them tomorrow. I also added some more parameters to help control when the measurement should fail, which I think could be useful, ie, if the measurement should fail on non-zero RC, if a timeout was hit, or a client-go related error

marseel reviewed (commented):
Thanks for the PR! 

Overall looks great, just a few minor nits.

learnitall commented:
@marseel I just pushed some changes. From the suggestion you had in regards to error handling, I was able to add a couple of new parameters which I think will be useful:

* `failOnCommandError`: If true, then the measurement will fail if one of the commands has a non-zero exit code.
* `failOnExecError`: If true, then the measurement will fail if the measurement fails to execute the command, such as if an error is returned from the client-go library.
* `failOnTimeout`: If true, then the measurement will fail if one of the commands takes longer than its configured timeout value.

It may be useful to play with these in some cases, such as for helping to address flakes in large-scale tests or during development.

I also adjusted the following:

* Measurements for stdout and stderr from commands now include `stdout` or `stderr` in their measurement name. This makes it easier to understand the output from CL2 if `--report-dir` is not given.
* Measurements names now have the timestamp up front, to make the output of `ls` in the report dir a bit easier to comb through.
* Fixed a bug where a nil pointer dereference can be forced by providing a bag config.

Here's an example config I used:

```yaml
{{$duration := "5s"}}
{{$namespaces := 1}}

name: generic-query
namespace:
  number: {{$namespaces}}
steps:
- name: Start measurements
  measurements:
  - Identifier: pp
    Method: PodPeriodicCommand
    Params:
      action: start
      labelSelector: "k8s-app=cilium"
      interval: "3s"
      container: "cilium-agent"
      limit: 1
      failOnCommandError: false
      failOnExecError: false
      failOnTimeout: false
      commands:
      - name: "do-error"
        command:
        - /bin/bash
        - idontexist
        timeout: 1s
      - name: "do-timeout"
        command:
        - sleep
        - 3
        timeout: 1s
      - name: "do-exec-error"
        command:
        - /bin/idontexist
        timeout: 1s
      - name: "do-success"
        command:
        - echo
        - "hello there"
        timeout: 1s
- name: Sleep
  measurements:
  - Identifier: sleep
    Method: Sleep
    Params:
      duration: {{$duration}}
- name: Gather measurements
  measurements:
  - Identifier: pp
    Method: PodPeriodicCommand
    Params:
      action: gather
```

And the output:

```
I1004 17:07:52.896177  282495 simple_test_executor.go:87] PodPeriodicCommand: [
  {
    "pod": "cilium-t9ms5",
    "namespace": "kube-system",
    "container": "cilium-agent",
    "commands": [
      {
        "exitCode": 127,
        "name": "do-error",
        "command": [
          "/bin/bash",
          "idontexist"
        ],
        "timeout": "1s",
        "hitTimeout": false,
        "startTime": "2023-10-04T17:07:47.856597479-06:00",
        "endTime": "2023-10-04T17:07:48.183532256-06:00",
        "execError": ""
      },
      {
        "exitCode": 0,
        "name": "do-exec-error",
        "command": [
          "/bin/idontexist"
        ],
        "timeout": "1s",
        "hitTimeout": false,
        "startTime": "2023-10-04T17:07:47.856664851-06:00",
        "endTime": "2023-10-04T17:07:48.193805375-06:00",
        "execError": "Internal error occurred: error executing command in container: failed to exec in container: failed to start exec \"dfef35d965723520d9b2d72a10e0a1a6c469c571267adfb33a5e4418bd45f6de\": OCI runtime exec failed: exec failed: unable to start container process: exec: \"/bin/idontexist\": stat /bin/idontexist: no such file or directory: unknown"
      },
      {
        "exitCode": 0,
        "name": "do-success",
        "command": [
          "echo",
          "hello there"
        ],
        "timeout": "1s",
        "hitTimeout": false,
        "startTime": "2023-10-04T17:07:47.856511677-06:00",
        "endTime": "2023-10-04T17:07:48.223041243-06:00",
        "execError": ""
      },
      {
        "exitCode": 0,
        "name": "do-timeout",
        "command": [
          "sleep",
          "3"
        ],
        "timeout": "1s",
        "hitTimeout": true,
        "startTime": "2023-10-04T17:07:47.856680418-06:00",
        "endTime": "2023-10-04T17:07:48.856804881-06:00",
        "execError": ""
      }
    ]
  },
  {
    "pod": "cilium-g7pqs",
    "namespace": "kube-system",
    "container": "cilium-agent",
    "commands": [
      {
        "exitCode": 0,
        "name": "do-success",
        "command": [
          "echo",
          "hello there"
        ],
        "timeout": "1s",
        "hitTimeout": false,
        "startTime": "2023-10-04T17:07:50.857177421-06:00",
        "endTime": "2023-10-04T17:07:51.128640741-06:00",
        "execError": ""
      },
      {
        "exitCode": 0,
        "name": "do-exec-error",
        "command": [
          "/bin/idontexist"
        ],
        "timeout": "1s",
        "hitTimeout": false,
        "startTime": "2023-10-04T17:07:50.857274317-06:00",
        "endTime": "2023-10-04T17:07:51.167844874-06:00",
        "execError": "Internal error occurred: error executing command in container: failed to exec in container: failed to start exec \"51181312bf60edd746eb6aaf0c0a7b0b8d51fb4abd7e7bd2248df22eae01160e\": OCI runtime exec failed: exec failed: unable to start container process: exec: \"/bin/idontexist\": stat /bin/idontexist: no such file or directory: unknown"
      },
      {
        "exitCode": 127,
        "name": "do-error",
        "command": [
          "/bin/bash",
          "idontexist"
        ],
        "timeout": "1s",
        "hitTimeout": false,
        "startTime": "2023-10-04T17:07:50.857300715-06:00",
        "endTime": "2023-10-04T17:07:51.195274682-06:00",
        "execError": ""
      },
      {
        "exitCode": 0,
        "name": "do-timeout",
        "command": [
          "sleep",
          "3"
        ],
        "timeout": "1s",
        "hitTimeout": true,
        "startTime": "2023-10-04T17:07:50.857282697-06:00",
        "endTime": "2023-10-04T17:07:51.857895274-06:00",
        "execError": ""
      }
    ]
  }
]
I1004 17:07:52.896217  282495 simple_test_executor.go:87] PodPeriodicCommand-2023-10-04T17:07:47-06:00-kube-system-cilium-t9ms5-do-error-stderr: /bin/bash: idontexist: No such file or directory
I1004 17:07:52.896230  282495 simple_test_executor.go:87] PodPeriodicCommand-2023-10-04T17:07:47-06:00-kube-system-cilium-t9ms5-do-success-stdout: hello there
I1004 17:07:52.896242  282495 simple_test_executor.go:87] PodPeriodicCommand-2023-10-04T17:07:50-06:00-kube-system-cilium-g7pqs-do-success-stdout: hello there
I1004 17:07:52.896247  282495 simple_test_executor.go:87] PodPeriodicCommand-2023-10-04T17:07:50-06:00-kube-system-cilium-g7pqs-do-error-stderr: /bin/bash: idontexist: No such file or directory
I1004 17:07:52.896253  282495 simple_test_executor.go:87] PodPeriodicCommand-stats: {
  "execs": 8,
  "execErrors": 2,
  "timeouts": 2,
  "nonZeroRCs": 2,
  "measurements": 6,
  "ticks": 2,
  "ticksNoPods": 0
}
```

marseel reviewed (commented):
Looks great, last two comments from me :)

marseel commented:
@dlapcevic could you take a look too? I think you might be interested in this measurement. Thanks!

learnitall commented on clusterloader2/pkg/measurement/common/pod_command.go:
```diff
@@ -0,0 +1,722 @@
+	isReady := false
+	for _, c := range pod.Status.Conditions {
+		if c.Type == v1.PodReady && c.Status == v1.ConditionTrue {
+			isReady = true
```
Nice catch!

learnitall commented on clusterloader2/pkg/util/util.go:
```diff
@@ -93,6 +95,11 @@ func GetStringArray(dict map[string]interface{}, key string) ([]string, error) {
 }
 
+// GetLabelSelector tries to return value from map parsed as labels.Selector type. If value doesn't exist, error is returned.
+func GetLabelSelector(dict map[string]interface{}, key string) (*labels.Selector, error) {
```
Done :+1:

learnitall commented:
Alright just pushed up those last changes you requested @marseel, should be good to go.

marseel commented:
/lgtm
/approve

/hold
for @dlapcevic lgtm

dlapcevic reviewed (approved):
lgtm

marseel commented:
/unhold
